From 802cd4ead5fd806794386dc57f97da79dfb3fd58 Mon Sep 17 00:00:00 2001
From: Ivan Kanakarakis <ivan.kanak@gmail.com>
Date: Sun, 18 Mar 2012 04:30:35 +0200
Subject: [PATCH] nmaster layout mode

---
 config.def.h |  3 +++
 monsterwm.c  | 32 +++++++++++++++++++++++---------
 2 files changed, 26 insertions(+), 9 deletions(-)

diff --git a/config.def.h b/config.def.h
index 4b2175d..50ade4f 100644
--- a/config.def.h
+++ b/config.def.h
@@ -26,6 +26,7 @@
 #define MINWSZ          50        /* minimum window size in pixels  */
 #define DEFAULT_DESKTOP 0         /* the desktop to focus initially */
 #define DESKTOPS        4         /* number of desktops - edit DESKTOPCHANGE keys to suit */
+#define NMASTER         1         /* number of windows in master area by default */
 
 /**
  * open applications to specified desktop with specified mode.
@@ -61,6 +62,8 @@ static Key keys[] = {
     {  MOD1|SHIFT,       XK_c,          killclient,        {NULL}},
     {  MOD1,             XK_j,          next_win,          {NULL}},
     {  MOD1,             XK_k,          prev_win,          {NULL}},
+    {  MOD1,             XK_d,          nmaster,           {.i = -1}},
+    {  MOD1,             XK_i,          nmaster,           {.i = +1}},
     {  MOD1,             XK_h,          resize_master,     {.i = -10}}, /* decrease size in px */
     {  MOD1,             XK_l,          resize_master,     {.i = +10}}, /* increase size in px */
     {  MOD1,             XK_o,          resize_stack,      {.i = -10}}, /* shrink   size in px */
diff --git a/monsterwm.c b/monsterwm.c
index 8d0b9d9..4b72830 100644
--- a/monsterwm.c
+++ b/monsterwm.c
@@ -88,6 +88,7 @@ static void move_up();
 static void moveresize(const Arg *arg);
 static void mousemotion(const Arg *arg);
 static void next_win();
+static void nmaster(const Arg *arg);
 static void prev_win();
 static void quit(const Arg *arg);
 static void resize_master(const Arg *arg);
@@ -131,9 +132,10 @@ typedef struct Client {
  * curr - the currently highlighted window
  * prev - the client that previously had focus
  * sbar - the visibility status of the panel/statusbar
+ * nm   - the number of windows in master area
  */
 typedef struct {
-    int mode, masz, sasz;
+    int mode, masz, sasz, nm;
     Client *head, *curr, *prev;
     Bool sbar;
 } Desktop;
@@ -887,6 +889,15 @@ void next_win(void) {
     if (d->curr && d->head->next) focus(d->curr->next ? d->curr->next:d->head, d);
 }
 
+/**
+ * increase or decrease the number
+ * of windows in the master area
+ */
+void nmaster(const Arg *arg) {
+    Desktop *d = &desktops[currdeskidx];
+    if ((d->nm += arg->i) >= 1) tile(d); else d->nm -= arg->i;
+}
+
 /**
  * get the previous client from the given
  * if no such client, return NULL
@@ -1026,7 +1037,7 @@ void setup(void) {
 
     /* initialize mode and panel visibility for each desktop */
     for (unsigned int d = 0; d < DESKTOPS; d++)
-        desktops[d] = (Desktop){ .mode = DEFAULT_MODE, .sbar = SHOW_PANEL };
+        desktops[d] = (Desktop){ .mode = DEFAULT_MODE, .sbar = SHOW_PANEL, .nm = NMASTER };
 
     /* get color for focused and unfocused client borders */
     win_focus = getcolor(FOCUS, screen);
@@ -1088,7 +1099,7 @@ void spawn(const Arg *arg) {
  */
 void stack(int x, int y, int w, int h, const Desktop *d) {
     Client *c = NULL, *t = NULL; Bool b = (d->mode == BSTACK);
-    int n = 0, p = 0, z = (b ? w:h), ma = (b ? h:w) * MASTER_SIZE + d->masz;
+    int n = 0, p = 0, z = (b ? w:h), ma = (b ? h:w) * MASTER_SIZE + d->masz, nm = d->nm;
 
     /* count stack windows and grab first non-floating, non-fullscreen window */
     for (t = d->head; t; t = t->next) if (!ISFFT(t)) { if (c) ++n; else c = t; }
@@ -1123,14 +1134,17 @@ void stack(int x, int y, int w, int h, const Desktop *d) {
      * and also, does not result in gaps created on the bottom of the screen.
      */
     if (c && !n) XMoveResizeWindow(dis, c->win, x, y, w - 2*BORDER_WIDTH, h - 2*BORDER_WIDTH);
-    if (!c || !n) return; else if (n > 1) { p = (z - d->sasz)%n + d->sasz; z = (z - d->sasz)/n; }
-
-    /* tile the first non-floating, non-fullscreen window to cover the master area */
-    if (b) XMoveResizeWindow(dis, c->win, x, y, w - 2*BORDER_WIDTH, ma - BORDER_WIDTH);
-    else   XMoveResizeWindow(dis, c->win, x, y, ma - BORDER_WIDTH, h - 2*BORDER_WIDTH);
+    if (!c || !n) return; else if (n - nm <= 0) nm = n;
+    else { p = (z - d->sasz)%((n -= nm-1)) + d->sasz; z = (z - d->sasz)/n; }
+
+    /* tile non-floating, non-fullscreen master windows to equally share the master area */
+    for (int i = 0; i < nm; i++) {
+        if (b) XMoveResizeWindow(dis, c->win, x + i * w/nm, y, w/nm - 2*BORDER_WIDTH, ma - BORDER_WIDTH);
+        else   XMoveResizeWindow(dis, c->win, x, y + i * h/nm, ma - BORDER_WIDTH, h/nm - 2*BORDER_WIDTH);
+        for (c = c->next; c && ISFFT(c); c = c->next);
+    }
 
     /* tile the next non-floating, non-fullscreen (and first) stack window adding p */
-    for (c = c->next; c && ISFFT(c); c = c->next);
     int cw = (b ? h:w) - 2*BORDER_WIDTH - ma, ch = z - BORDER_WIDTH;
     if (b) XMoveResizeWindow(dis, c->win, x, y += ma, ch - BORDER_WIDTH + p, cw);
     else   XMoveResizeWindow(dis, c->win, x += ma, y, cw, ch - BORDER_WIDTH + p);
-- 
2.52.0

