From f3155de43dce108503ed4c6754f337bde63fa09a Mon Sep 17 00:00:00 2001
From: Ivan Kanakarakis <ivan.kanak@gmail.com>
Date: Fri, 6 Jan 2012 03:14:53 +0200
Subject: [PATCH] showhide() function toggles visibility of all windows on all
 desktops

---
 config.def.h |  1 +
 monsterwm.c  | 18 ++++++++++++++----
 2 files changed, 15 insertions(+), 4 deletions(-)

diff --git a/config.def.h b/config.def.h
index 4b2175d..480da4e 100644
--- a/config.def.h
+++ b/config.def.h
@@ -57,6 +57,7 @@ static const char *menucmd[] = { "dmenu_run", NULL };
 static Key keys[] = {
     /* modifier          key            function           argument */
     {  MOD1,             XK_b,          togglepanel,       {NULL}},
+    {  MOD4,             XK_s,          showhide,          {NULL}},
     {  MOD1,             XK_BackSpace,  focusurgent,       {NULL}},
     {  MOD1|SHIFT,       XK_c,          killclient,        {NULL}},
     {  MOD1,             XK_j,          next_win,          {NULL}},
diff --git a/monsterwm.c b/monsterwm.c
index 8d0b9d9..070b839 100644
--- a/monsterwm.c
+++ b/monsterwm.c
@@ -94,6 +94,7 @@ static void resize_master(const Arg *arg);
 static void resize_stack(const Arg *arg);
 static void rotate(const Arg *arg);
 static void rotate_filled(const Arg *arg);
+static void showhide();
 static void spawn(const Arg *arg);
 static void swap_master();
 static void switch_mode(const Arg *arg);
@@ -184,7 +185,7 @@ static int xerrorstart(Display *dis, XErrorEvent *ee);
  * desktops     - array of managed desktops
  * currdeskidx  - which desktop is currently active
  */
-static Bool running = True;
+static Bool running = True, show = True;
 static int wh, ww, currdeskidx, prevdeskidx, retval;
 static unsigned int numlockmask, win_unfocus, win_focus;
 static Display *dis;
@@ -271,8 +272,8 @@ void buttonpress(XEvent *e) {
 void change_desktop(const Arg *arg) {
     if (arg->i == currdeskidx || arg->i < 0 || arg->i >= DESKTOPS) return;
     Desktop *d = &desktops[(prevdeskidx = currdeskidx)], *n = &desktops[(currdeskidx = arg->i)];
-    if (n->curr) XMapWindow(dis, n->curr->win);
-    for (Client *c = n->head; c; c = c->next) XMapWindow(dis, c->win);
+    if (n->curr && show) XMapWindow(dis, n->curr->win);
+    for (Client *c = n->head; show && c; c = c->next) XMapWindow(dis, c->win);
     XChangeWindowAttributes(dis, root, CWEventMask, &(XSetWindowAttributes){.do_not_propagate_mask = SubstructureNotifyMask});
     for (Client *c = d->head; c; c = c->next) if (c != d->curr) XUnmapWindow(dis, c->win);
     if (d->curr) XUnmapWindow(dis, d->curr->win);
@@ -725,7 +726,7 @@ void maprequest(XEvent *e) {
         setfullscreen(c, d, (*(Atom *)state == netatoms[NET_FULLSCREEN]));
     if (state) XFree(state);
 
-    if (currdeskidx == newdsk) { if (!ISFFT(c)) tile(d); XMapWindow(dis, c->win); }
+    if (currdeskidx == newdsk) { if (!ISFFT(c)) tile(d); if (show) XMapWindow(dis, c->win); }
     else if (follow) change_desktop(&(Arg){.i = newdsk});
     focus(c, d);
 
@@ -1182,6 +1183,15 @@ void tile(Desktop *d) {
                                                   ww, wh + (d->sbar ? 0:PANEL_HEIGHT), d);
 }
 
+/**
+ * toggle visibility of all windows in all desktops
+ */
+void showhide(void) {
+    if ((show = !show)) tile(&desktops[currdeskidx]);
+    else for (Client *c = desktops[currdeskidx].head; c; c = c->next)
+        XMoveWindow(dis, c->win, - 2 * ww, 0);
+}
+
 /**
  * toggle visibility state of the panel/bar
  */
-- 
2.52.0

